### ISSUE TITLE
db:sync Override Hook and File Paths Question 

### ISSUE THREAD

**Issue by mgamshad** [acquia/blt#2300]
My system information:
* Operating system type: macOS Sierra
* Operating system version: 10.12.4
* BLT version: 8.9.x

Question: I am experimenting with overriding the `db:sync` hook to use `drush sqlq` instead, and have a class created using `blt examples:init` which is functioning correctly. I am also leveraging a setting in the `project.local.yml` to specify a path to a `.sql` file so that the override command can use that.  

However, I noticed though that the path picked up from the `project.local.yml` file is being treated as a path inside the VM.  Is there a way to have that path be treated to be in the host environment, or a logic where I can check either in host or the VM?

Thanks,
Mohsen

**Comment by danepowell** [comment]
I assume you're referring to `sync:db`, since there's no such command as `db:sync`.

You also don't specify exactly which version of BLT you are using, which is problematic, because this behavior has changed from version to version.

Whether a command is executed inside or outside of the VM is governed by the `@executeInDrupalVm` annotation.

In version 8.9.10, BLT should execute the `sync` and `sync:db` commands outside of the VM, because there is no such annotation: https://github.com/acquia/blt/blob/8bf4845e778ca52c95e38cb1a68c63c1612b1e9f/src/Robo/Commands/Sync/DbCommand.php#L53

In version 9.1.x, the commands are executed inside the VM:
https://github.com/acquia/blt/blob/b9ad03faca3c2192bb02a9a6f415109ea08c0a87/src/Robo/Commands/Sync/DbCommand.php#L57

**Comment by mgamshad** [comment]
Sorry for some missing details.  Yes, I was referring to `sync:db` command, and the exact version of BLT is 8.9.11.  (I had listed 8.9.x, as we are using 8.9.x-dev in our composer definition).

Having said that, 8.9.11 seems to be behaving as it's executing the commands inside the VM even though the annotation is not there.  Try overriding the `sync:db` with a hook to load the DB from a local .sql file.  Something similar to:
```
 $dbFile = $this->getConfigValue('local-db-override.db-file');

 $this->say("Local DB Override enabled.");

 $this->say("Local DB Override file = $dbFile");

 $this->invokeCommand('setup:settings');

 $local_alias = '@' . $this->getConfigValue('drush.aliases.local');

 $task = $this->taskDrush()
     ->drush('cache-clear drush')
     ->drush('sql-drop')
     ->drush('sql-query')
     ->arg($local_alias)
     ->option('file', $dbFile)
     ->assume(TRUE);

 $task->drush('cache-clear drush');

 $result = $task->run();

 if (!$result->wasSuccessful()) {
     $this->logger->error("Could not complete DB sync with local override");
     throw new BltException("Could not sync database.");
 }

 return $result;
```

and have the path to the DB file be something like: `/tmp/database.sql`.  

After placing the above hook in a `*hook.php` file in my project's `blt/src/Hooks` folder, running the command `blt sync` results in an error where the message indicates the file `/tmp/database.sql` is not found.  However, if I `vagrant ssh` into my box and copy that file to the VM's `/tmp` folder, the command executes fine.

**Comment by danepowell** [comment]
I think I understand now, this is a problem with how you are using Drush. It has nothing to do with the BLT sync command. You are effectively running this command:

`drush @my-project.local sqlq --file=dbfile.sql`

Because you are using the alias _@my-project.local_, Drush is going to execute this command inside the VM.

It sounds like you are duplicating built-in BLT functionality. BLT natively supports importing a database dump instead of running site-install. Just set `setup.strategy` and `setup.dump-file`. For example: https://github.com/acquia/blt/blob/8.9.x/template/blt/ci.yml

In this case, the strategy you'd want is "import". If you're curious, this is how BLT would actually do the import: https://github.com/acquia/blt/blob/8.9.x/src/Robo/Commands/Setup/ImportCommand.php

There's a lot of different ways you can set these config values: project-wide, per-site (on multisite), per-environment (i.e. only on local or CI), or on the command line.

If this works for you and there's some way we can make it more obvious/usable, please let me know.

can't you use setup strategy from ci.yml?

**Comment by mgamshad** [comment]
So, I would rather prefer an approach that `blt` supports out of the box.  I tried the following in the ci.yml file:
```
setup.strategy: import
setup.dump-file: /tmp/database.sql
```
And I did not see any references to that file in the logs when `blt setup` or `blt sync` were running, which made me doubt if it was being used.  Is there something else that needs to be set, or a different command that needs to be invoked?

Here is the main requirement for what we are trying to do: in some environments ssh port is closed so developers can't execute commands that retrieve the database from a remote system.  For those cases we need to have a way to let the `blt` commands know that instead of accessing the remote system for the database, it should load it from a local file.  But I'd like that to be within the framework of a common set of commands that are preformed by all the developers.  So, if setting up a local sandbox involves executing the following set of commands:
```
composer install
blt vm
blt setup
blt sync
```
I would like **all** the developers to use the same set of commands, where some would perhaps specify a config entry that would select the DB file from a local versus a sync against a remote system, but they are all still using the same commands.  Given those requirements, how would I configure the project settings to accomplish that?  If this is possible within `blt`, and to answer your question, some concrete documentation would be helpful.  For example, I couldn't find anything in the docs on `ci.yml` at all!!

**Comment by danepowell** [comment]
`ci.yml`, as the name implies, is only used in CI environments. If you want `blt setup` to always use a dump file, then you should define the `setup.strategy` and `setup.dump-file` keys in `blt/project.yml`, not `ci.yml`.

**Comment by mgamshad** [comment]
hmm, so I tried adding those two to the `blt/project.yml` file and still see no logs indicating that it's picking up the file, instead `blt sync:refresh` is still invoking `drush sql-sync ...`???

**Comment by danepowell** [comment]
Run `blt setup`, not `blt sync:refresh`.

You might benefit from examining the code behind these commands.

https://github.com/acquia/blt/blob/190d0d610cb2fa302037b3cb01899ab06b1aeb73/src/Robo/Commands/Sync/RefreshCommand.php#L60

https://github.com/acquia/blt/blob/190d0d610cb2fa302037b3cb01899ab06b1aeb73/src/Robo/Commands/Setup/AllCommand.php#L15

Note that `blt setup` is a more generic wrapper command that will delegate to `sync:refresh` or `setup:import` based on `setup.strategy`.

**Comment by mgamshad** [comment]
Running `blt setup` does not indicate in any way that it's retrieving the file listed in the configs?  Also, `blt setup` does not seem to be the only command we need to run, since do so just performs the install and we don't get any of the live site's content or DB.  We get that only after running `blt sync:refresh`.  So, for now it doesn't look like I am having any luck executing the functionality that you are claiming to be in `blt`, unless I am missing something.

**Comment by danepowell** [comment]
Now I'm confused about what you are trying to accomplish. If you want to "get any of the live site's content or data", you should _either_ invoke `blt sync:refresh` or set `setup.strategy` to `sync` and invoke `blt setup`. If you want to import a static database dump from disk, you should set `setup.strategy` to `import` and invoke `blt setup`.

If that's not working, please paste the contents of your `project.yml` and complete log output from `blt setup`.

**Comment by grasmash** [comment]
@mgamshad Should this issue remain active?

**Comment by mgamshad** [comment]
@danepowell  I did exactly as you stated, i.e. set the `setup.strategy` to `import` and ran `blt setup`, and what I reported was that the import from the file set in `setup.dump-file` was not happening.  And again, to repeat, the main reason we are doing this is that we are behind a corporate firewall and can't SSH into any remote machines.

@grasmash I'd like to get my head around the questions though.  Thanks.
